be/src/exec/parquet_builder.cpp

```diff
@@ -0,0 +1,351 @@
+        break;
+    }
+    case TYPE_DOUBLE: {
+        parquet_data_type = parquet::Type::DOUBLE;
+        break;
+    }
+    case TYPE_CHAR:
+    case TYPE_VARCHAR:
+    case TYPE_DECIMAL:
+    case TYPE_DECIMAL32:
+    case TYPE_DECIMAL64:
+    case TYPE_DECIMALV2: {
+        parquet_data_type = parquet::Type::BYTE_ARRAY;
+        break;
+    }
+    default:
+        parquet_data_type = parquet::Type::UNDEFINED;
+    }
+}
+
+void ParquetBuildHelper::build_parquet_repetition_type(parquet::Repetition::type& parquet_repetition_type,
+                                                       const bool is_nullable) {
+    parquet_repetition_type = is_nullable ? parquet::Repetition::OPTIONAL : parquet::Repetition::REQUIRED;
+}
+
+void ParquetBuildHelper::build_compression_type(parquet::WriterProperties::Builder& builder,
+                                                const TCompressionType::type& compression_type) {
+    switch (compression_type) {
+    case TCompressionType::SNAPPY: {
+        builder.compression(parquet::Compression::SNAPPY);
+        break;
+    }
+    case TCompressionType::GZIP: {
+        builder.compression(parquet::Compression::GZIP);
+        break;
+    }
+    case TCompressionType::BROTLI: {
+        builder.compression(parquet::Compression::BROTLI);
+        break;
+    }
+    case TCompressionType::ZSTD: {
+        builder.compression(parquet::Compression::ZSTD);
+        break;
+    }
+    case TCompressionType::LZ4: {
+        builder.compression(parquet::Compression::LZ4);
+        break;
+    }
+    case TCompressionType::LZO: {
+        builder.compression(parquet::Compression::LZO);
+        break;
+    }
+    case TCompressionType::BZIP2: {
+        builder.compression(parquet::Compression::BZ2);
+        break;
+    }
+    case TCompressionType::DEFAULT_COMPRESSION: {
+        builder.compression(parquet::Compression::UNCOMPRESSED);
+        break;
+    }
+    default:
+        builder.compression(parquet::Compression::UNCOMPRESSED);
+    }
+}
+
+void ParquetBuilder::_generate_rg_writer() {
+    if (_rg_writer == nullptr) {
+        _rg_writer = _file_writer->AppendBufferedRowGroup();
+        _cur_written_rows = 0;
+    }
+}
+
+#define DISPATCH_PARQUET_NUMERIC_WRITER(WRITER, COLUMN_TYPE, NATIVE_TYPE)                                         \
+    ParquetBuilder::_generate_rg_writer();                                                                        \
+    parquet::WRITER* col_writer = static_cast<parquet::WRITER*>(_rg_writer->column(i));                           \
+    col_writer->WriteBatch(                                                                                       \
+            num_rows, nullable ? def_level.data() : nullptr, nullptr,                                             \
+            reinterpret_cast<const NATIVE_TYPE*>(down_cast<const COLUMN_TYPE*>(data_column)->get_data().data())); \
+    _buffered_values_estimate[i] = col_writer->EstimatedBufferedValueBytes();
```

Comment:
why not calculate it until we need it

Reply:
that would be a lot of duplicate codes.

Reply:
I don't understand why there will be duplicate codes. Could you elaborate it a little bit

Reply:
I need to cast column_writer to TypedColumnWriter<Type> to use the EstimatedBufferedValueBytes() interface. so I also need to get accurate every column type in here using duplicate codes.